File: rs/tests/test_glob.rs
mod common;

use vost::*;

fn store_with_glob_files(dir: &std::path::Path) -> (GitStore, Fs) {
    let store = common::create_store(dir, "main");
    let fs = store.branches().get("main").unwrap();

    let mut batch = fs.batch(Default::default());
    batch.write("readme.txt", b"readme").unwrap();
    batch.write("notes.txt", b"notes").unwrap();
    batch.write("data.csv", b"data").unwrap();
    batch.write(".hidden", b"hidden").unwrap();
    batch.write("src/main.py", b"main").unwrap();
    batch.write("src/lib.py", b"lib").unwrap();
    batch.write("src/util.rs", b"util").unwrap();
    batch.write("src/deep/mod.py", b"mod").unwrap();
    batch.write("src/deep/nested/core.py", b"core").unwrap();
    batch.write("docs/guide.md", b"guide").unwrap();
    batch.write("docs/api.md", b"api").unwrap();
    batch.commit().unwrap();

    let fs = store.branches().get("main").unwrap();
    (store, fs)
}

// ---------------------------------------------------------------------------
// Star (*)
// ---------------------------------------------------------------------------

#[test]
fn glob_star_txt() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("*.txt").unwrap();
    assert_eq!(matches, vec!["notes.txt", "readme.txt"]);
}

#[test]
fn glob_star_excludes_dotfiles() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("*").unwrap();
    // Should not include .hidden
    assert!(!matches.contains(&".hidden".to_string()));
    // But should include regular files
    assert!(matches.contains(&"data.csv".to_string()));
}

#[test]
fn glob_dotstar_matches_dotfiles() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob(".*").unwrap();
    assert_eq!(matches, vec![".hidden"]);
}

#[test]
fn glob_star_in_subdir() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/*.py").unwrap();
    assert_eq!(matches, vec!["src/lib.py", "src/main.py"]);
}

#[test]
fn glob_extension_filter() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("*.csv").unwrap();
    assert_eq!(matches, vec!["data.csv"]);
}

// ---------------------------------------------------------------------------
// Question mark (?)
// ---------------------------------------------------------------------------

#[test]
fn glob_question_mark() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write("abc.md", b"a").unwrap();
    batch.write("xyz.md", b"b").unwrap();
    batch.write("ab.md", b"c").unwrap();
    batch.commit().unwrap();

    let fs = store.branches().get("main").unwrap();
    let matches = fs.glob("???.md").unwrap();
    assert_eq!(matches, vec!["abc.md", "xyz.md"]);
}

// ---------------------------------------------------------------------------
// Mixed patterns
// ---------------------------------------------------------------------------

#[test]
fn glob_literal_then_glob() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("docs/*.md").unwrap();
    assert_eq!(matches, vec!["docs/api.md", "docs/guide.md"]);
}

#[test]
fn glob_glob_then_literal() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/util.rs").unwrap();
    assert_eq!(matches, vec!["src/util.rs"]);
}

// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------

#[test]
fn glob_no_matches() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("*.xyz").unwrap();
    assert!(matches.is_empty());
}

#[test]
fn glob_literal_path() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("readme.txt").unwrap();
    assert_eq!(matches, vec!["readme.txt"]);
}

#[test]
fn glob_literal_missing() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("nonexistent.txt").unwrap();
    assert!(matches.is_empty());
}

#[test]
fn glob_results_sorted() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("*.txt").unwrap();
    let mut sorted = matches.clone();
    sorted.sort();
    assert_eq!(matches, sorted);
}

// ---------------------------------------------------------------------------
// Double star (**)
// ---------------------------------------------------------------------------

#[test]
fn glob_doublestar_recursive() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("**/*.py").unwrap();
    assert!(matches.contains(&"src/main.py".to_string()));
    assert!(matches.contains(&"src/lib.py".to_string()));
    assert!(matches.contains(&"src/deep/mod.py".to_string()));
    assert!(matches.contains(&"src/deep/nested/core.py".to_string()));
}

#[test]
fn glob_doublestar_prefix() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/**/*.py").unwrap();
    assert!(matches.contains(&"src/main.py".to_string()));
    assert!(matches.contains(&"src/lib.py".to_string()));
    assert!(matches.contains(&"src/deep/mod.py".to_string()));
    assert!(matches.contains(&"src/deep/nested/core.py".to_string()));
    // Should not include non-src files
    assert!(!matches.iter().any(|m| m.starts_with("docs/")));
}

#[test]
fn glob_doublestar_deep() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/**/core.py").unwrap();
    assert_eq!(matches, vec!["src/deep/nested/core.py"]);
}

#[test]
fn glob_doublestar_no_dotfiles() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write(".dotdir/file.txt", b"a").unwrap();
    batch.write("normal/file.txt", b"b").unwrap();
    batch.commit().unwrap();

    let fs = store.branches().get("main").unwrap();
    let matches = fs.glob("**/*.txt").unwrap();
    // .dotdir should be skipped
    assert!(!matches.iter().any(|m| m.contains(".dotdir")));
    assert!(matches.contains(&"normal/file.txt".to_string()));
}

#[test]
fn glob_doublestar_no_duplicates() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("**/*.py").unwrap();
    let mut unique = matches.clone();
    unique.sort();
    unique.dedup();
    assert_eq!(matches.len(), unique.len());
}

#[test]
fn glob_doublestar_empty_repo() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let matches = fs.glob("**/*.txt").unwrap();
    assert!(matches.is_empty());
}

#[test]
fn glob_doublestar_sorted() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("**/*.py").unwrap();
    let mut sorted = matches.clone();
    sorted.sort();
    assert_eq!(matches, sorted);
}

// ---------------------------------------------------------------------------
// iglob (if available — same as glob, just iterator)
// ---------------------------------------------------------------------------

#[test]
fn glob_star_rs_extension() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/*.rs").unwrap();
    assert_eq!(matches, vec!["src/util.rs"]);
}

#[test]
fn glob_doublestar_all_md() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("**/*.md").unwrap();
    assert!(matches.contains(&"docs/api.md".to_string()));
    assert!(matches.contains(&"docs/guide.md".to_string()));
    assert_eq!(matches.len(), 2);
}

#[test]
fn glob_doublestar_mixed_extensions() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let py_matches = fs.glob("**/*.py").unwrap();
    let rs_matches = fs.glob("**/*.rs").unwrap();
    assert_eq!(py_matches.len(), 4); // main.py, lib.py, mod.py, core.py
    assert_eq!(rs_matches.len(), 1); // util.rs
}

#[test]
fn glob_star_empty_repo() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let matches = fs.glob("*.txt").unwrap();
    assert!(matches.is_empty());
}

#[test]
fn glob_doublestar_all_files() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("**/*").unwrap();
    // Should get all non-dotfiles
    assert!(matches.contains(&"readme.txt".to_string()));
    assert!(matches.contains(&"src/main.py".to_string()));
    assert!(matches.contains(&"src/deep/nested/core.py".to_string()));
    assert!(!matches.contains(&".hidden".to_string()));
}

#[test]
fn glob_question_mark_in_subdir() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write("dir/ab.txt", b"a").unwrap();
    batch.write("dir/cd.txt", b"b").unwrap();
    batch.write("dir/abc.txt", b"c").unwrap();
    batch.commit().unwrap();

    let fs = store.branches().get("main").unwrap();
    let matches = fs.glob("dir/??.txt").unwrap();
    assert_eq!(matches, vec!["dir/ab.txt", "dir/cd.txt"]);
}

#[test]
fn glob_doublestar_star_at_end() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_glob_files(dir.path());
    let matches = fs.glob("src/**/*").unwrap();
    // Should include all files under src/
    assert!(matches.contains(&"src/main.py".to_string()));
    assert!(matches.contains(&"src/lib.py".to_string()));
    assert!(matches.contains(&"src/util.rs".to_string()));
    assert!(matches.contains(&"src/deep/mod.py".to_string()));
    assert!(matches.contains(&"src/deep/nested/core.py".to_string()));
}

File: cpp/tests/test_glob.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <algorithm>
#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_gtest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

/// Helper: create a store with some files for glob testing.
static vost::Fs make_glob_fixture(vost::GitStore& store) {
    auto snap = store.branches().get("main");
    // Create a file structure:
    // readme.txt
    // src/main.cpp
    // src/util.cpp
    // src/lib/helper.h
    // src/lib/helper.cpp
    // .hidden
    // .config/settings.json
    // docs/guide.md
    // docs/api.md
    auto mk = [](const std::string& s) -> vost::WriteEntry {
        return vost::WriteEntry::from_text(s);
    };
    std::vector<std::pair<std::string, vost::WriteEntry>> writes = {
        {"readme.txt",            mk("README")},
        {"src/main.cpp",          mk("main")},
        {"src/util.cpp",          mk("util")},
        {"src/lib/helper.h",      mk("h")},
        {"src/lib/helper.cpp",    mk("c")},
        {".hidden",               mk("h")},
        {".config/settings.json", mk("s")},
        {"docs/guide.md",         mk("g")},
        {"docs/api.md",           mk("a")},
    };
    return snap.apply(writes);
}

// ---------------------------------------------------------------------------
// Basic glob tests
// ---------------------------------------------------------------------------

TEST_CASE("Glob: *.txt matches top-level txt files", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("*.txt");
    REQUIRE(results.size() == 1);
    CHECK(results[0] == "readme.txt");

    fs::remove_all(path);
}

TEST_CASE("Glob: *.cpp matches nothing at top level", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("*.cpp");
    CHECK(results.empty());

    fs::remove_all(path);
}

TEST_CASE("Glob: src/*.cpp matches cpp files in src/", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("src/*.cpp");
    REQUIRE(results.size() == 2);
    CHECK(results[0] == "src/main.cpp");
    CHECK(results[1] == "src/util.cpp");

    fs::remove_all(path);
}

TEST_CASE("Glob: src/lib/* matches all files in src/lib/", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("src/lib/*");
    REQUIRE(results.size() == 2);
    CHECK(results[0] == "src/lib/helper.cpp");
    CHECK(results[1] == "src/lib/helper.h");

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// ? wildcard
// ---------------------------------------------------------------------------

TEST_CASE("Glob: ? matches single character", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("docs/???.md");
    REQUIRE(results.size() == 1);
    CHECK(results[0] == "docs/api.md");

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// ** recursive
// ---------------------------------------------------------------------------

TEST_CASE("Glob: **/*.cpp matches all cpp files recursively", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("**/*.cpp");
    REQUIRE(results.size() == 3);
    // sorted: src/lib/helper.cpp, src/main.cpp, src/util.cpp
    CHECK(results[0] == "src/lib/helper.cpp");
    CHECK(results[1] == "src/main.cpp");
    CHECK(results[2] == "src/util.cpp");

    fs::remove_all(path);
}

TEST_CASE("Glob: **/*.md matches all md files recursively", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("**/*.md");
    REQUIRE(results.size() == 2);
    CHECK(results[0] == "docs/api.md");
    CHECK(results[1] == "docs/guide.md");

    fs::remove_all(path);
}

TEST_CASE("Glob: ** matches all files recursively", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    // ** alone should match everything (except dotfiles)
    // But ** only matches directories — needs a trailing wildcard segment
    // Actually, per the Rust implementation, ** with no following segment
    // matches zero levels (no-op), so this should be empty.
    // Let me use **/* instead.
    auto results = snap.glob("**/*");
    // Should match all non-dotfile files: readme.txt, src/main.cpp,
    // src/util.cpp, src/lib/helper.h, src/lib/helper.cpp,
    // docs/guide.md, docs/api.md
    CHECK(results.size() == 7);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// Dotfile handling
// ---------------------------------------------------------------------------

TEST_CASE("Glob: * does not match dotfiles", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("*");
    // Should match readme.txt only (not .hidden, not .config dir)
    CHECK(results.size() == 1);
    CHECK(results[0] == "readme.txt");

    fs::remove_all(path);
}

TEST_CASE("Glob: .* matches dotfiles explicitly", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob(".*");
    REQUIRE(results.size() == 1);
    CHECK(results[0] == ".hidden");

    fs::remove_all(path);
}

TEST_CASE("Glob: .config/* matches files in dotfile directory", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob(".config/*");
    REQUIRE(results.size() == 1);
    CHECK(results[0] == ".config/settings.json");

    fs::remove_all(path);
}

TEST_CASE("Glob: ** skips dotfile directories", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("**/*.json");
    // .config/settings.json should NOT be found because ** skips dotfile dirs
    CHECK(results.empty());

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// iglob returns unsorted
// ---------------------------------------------------------------------------

TEST_CASE("Glob: iglob returns same results as glob (possibly unsorted)", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto sorted = snap.glob("**/*.cpp");
    auto unsorted = snap.iglob("**/*.cpp");

    // Same elements
    auto sorted_copy = unsorted;
    std::sort(sorted_copy.begin(), sorted_copy.end());
    CHECK(sorted == sorted_copy);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------

TEST_CASE("Glob: empty pattern returns nothing", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("");
    CHECK(results.empty());

    fs::remove_all(path);
}

TEST_CASE("Glob: literal path matches exact file", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("readme.txt");
    REQUIRE(results.size() == 1);
    CHECK(results[0] == "readme.txt");

    fs::remove_all(path);
}

TEST_CASE("Glob: no matches returns empty", "[glob]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = make_glob_fixture(store);

    auto results = snap.glob("*.xyz");
    CHECK(results.empty());

    fs::remove_all(path);
}
